kernel/src/main/java/org/kframework/compile/ConstantFolding.java

```diff
@@ -0,0 +1,723 @@
+// Copyright (c) 2021 K Team. All Rights Reserved.
+package org.kframework.compile;
+
+import org.apache.commons.lang3.StringUtils;
+import org.kframework.attributes.Att;
+import org.kframework.builtin.Hooks;
+import org.kframework.definition.Module;
+import org.kframework.definition.Rule;
+import org.kframework.definition.Sentence;
+import org.kframework.kore.K;
+import org.kframework.kore.KApply;
+import org.kframework.kore.KToken;
+import org.kframework.kore.Sort;
+import org.kframework.kore.TransformK;
+import org.kframework.utils.errorsystem.KEMException;
+import org.kframework.utils.StringUtil;
+import org.kframework.mpfr.BigFloat;
+import org.kframework.mpfr.BinaryMathContext;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.math.BigInteger;
+import java.math.RoundingMode;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.kframework.Collections.*;
+import static org.kframework.kore.KORE.*;
+import static org.kframework.definition.Constructors.*;
+
+public class ConstantFolding {
+
+  private static final List<String> hookNamespaces = Arrays.asList(Hooks.BOOL, Hooks.FLOAT, Hooks.INT, Hooks.STRING);
+
+  private K loc;
+
+  void setLoc(K loc) {
+    this.loc = loc;
+  }
+
+  public Sentence fold(Module module, Sentence sentence) {
+    if (sentence instanceof Rule) {
+      Rule r = (Rule)sentence;
+      return Rule(fold(module, r.body(), true), fold(module, r.requires(), false), fold(module, r.ensures(), false), r.att());
+    }
+    return sentence;
+  }
+
+  public K fold(Module module, K body, boolean isBody) {
+    return new RewriteAwareTransformer(isBody) {
+      @Override
+      public K apply(KApply k) {
+        if (isLHS() || !isRHS()) {
+          return super.apply(k);
+        }
+        Att att = module.attributesFor().get(k.klabel()).getOrElse(() -> Att());
+        if (att.contains(Att.HOOK()) && !att.contains(Att.IMPURE())) {
+          String hook = att.get(Att.HOOK());
+          if (hookNamespaces.stream().anyMatch(ns -> hook.startsWith(ns + "."))) {
+            List<K> args = new ArrayList<>();
+            for (K arg : k.items()) {
+              K expanded = apply(arg);
+              if (!(expanded instanceof KToken)) {
+                return super.apply(k);
+              }
+              args.add(expanded);
+            }
+            try {
+              loc = k;
+              return doFolding(hook, args, module.productionsFor().apply(k.klabel().head()).head().substitute(k.klabel().params()).sort(), module);
+            } catch (NoSuchMethodException e) {
+              throw KEMException.internalError("Missing constant-folding implementation for hook " + hook, e);
+            }
```

Comment:
Rather than throw an error, I think it should be a warning. Do we assume that every hooked function in the above namespaces _must_ have constant-folding enabled for it?

Reply:
I mean, I'd rather it be an error so that we notice immediately that one is missing and add it prior to the code being deployed.